File: calculation.m
function [per_result] = calculation(model_n, fbasol_fin, randval_fin, method)

% calculation finds the optimal values in flux distribution profiles following perturbation that
% are significantly different (>0.05 0r >0.000001) from the original FBA/FVA flux distribution
% profiles.

%USAGE: 
%    [per_result] = calculation(model, fbasol_fin, randval_fin)

%INPUT

%model: metabolic model in SBML format

%fbasol_fin: flux distribution profiles following perturbation

%randval_fin: random values at which variable reactions were fixed

%method: 'fba' or 'fva', to determine the method of use to obtain flux

%OUTPUT:

%per_result: significantly different flux values in a sorted format

%Example:

%calculation [per_result] = calculation('mut-chem.xml','fbasol_fin.dat','randval_fin.dat','fba')

% Authors:

% Seyed Babak Loghmani

% Last updated: August 2021

model = readCbModel(model_n);
load(fbasol_fin)
load(randval_fin)

%FVA
[minFluxF1, maxFluxF1, optsol, ret, fbasol, fvamin, fvamax, statussolmin, statussolmax] = fastFVA(model);

if method == 'fva'
    opt_sol = fbasol;
elseif method == 'fba'
    sol_dist = optimizeCbModel(model);
    opt_sol = sol_dist.v;
end

%finding reactions with fva intervals larger that 0.000001
fva_n = maxFluxF1 - minFluxF1;
r=fva_n > 0.000001;
fva_n_f=find(r);

n = numel(model.rxns);
opt_som = [];
sols = [];
per_result = [];
%opt_sol = fbasol(1:n);
opt_sol_abs = abs(opt_sol);

% defining the 5% tolerance
tol = 5*opt_sol_abs/100;
s = size(randval_fin);
f = 0;
rxn_n = numel(model.rxns);


%finding and sorting significant flux changes, assigning their indices and random values in which they
%got fixed

num_f_p = numel(fva_n_f)/50;
num_f = ceil(num_f_p);
rg = 1:num_f;
c=1;
file_names = [];

for i = 1:numel(fva_n_f)
    k = i-1;
    
    l = (k*n) + 1;%lowe3r end of the data set
    h = (i*n);%higher end of the data set

    
if i == rg(c)*50%counter to save the data regarding ecery 50 reactions at once
    f = 1
    c = c+1;
elseif c == max(rg)
    
    if i == numel(fva_n_f)
       f = 1
    end 
end


if i ==1
    
    sols = fbasol_fin(1:rxn_n,1:10);%extracting the data for the first reaction
else
    sols = fbasol_fin(l:h,1:10);%extracting the data for the other reactions
end

for j = 1:10

    dif = opt_sol - sols(:,j);
    dif_abs = abs(dif);
    for k = 1:rxn_n
        
        %finding changes in flux values larger than 0.000001 for
        %reactions whose original value was 0
        
        if tol(k)==0
           if dif_abs(k) > 0.000001
              opt_som = [i,j, k, sols(k,j), opt_sol(k),tol(k), randval_fin(i,j)];
              per_result = [per_result; opt_som];
            
           end
           
        else
            
           %finding changes in flux values larger than 5% tolernace
           
           if dif_abs(k) > tol(k)
               
               %saving significantly different flux values in per_result
               
               opt_som = [i,j, k, sols(k,j), opt_sol(k),tol(k), randval_fin(i,j)];
               per_result = [per_result; opt_som];
           end

        end
        
    end
end

%saving sorted flux values in multiple files for the sake of memory usage

if f ==1
filename = sprintf('%s%d','per_result_new',i, '.dat');
save(filename,'per_result','-ascii','-double');

f_str = string(filename)
file_names = [file_names, f_str];

f=0;
clear per_result

per_result =[];

    end
end


%merging all the files into one comprehensive file
for i = 1:numel(file_names);
    
    file = load(file_names(i));
    per_result = [per_result;file];
end
save('per_result','per_result','-ascii','-double');
            
        

File: calculation.py
import os
import cbmpy as cbm
import numpy as np
import random


def calculation(model, pert_result, method):
    
    """calculates the number of significantly change flux values 
       and saves the results in `final.csv` which can be used for 
       statistical analysis either in Matlab or python.

       :param model: the model filename to be used (is assumed to 
                     be in the current directory)
       :type model: str
       
       :param pert_result: csv file of pertubations obtained by running `perturb`
       :type pert_result: str 
    
       :param method: the method to be used either 'fba' or 'fva'
       :type method: str
    
        Example:
    
        calculation(model = 'mut-chem.xml', pert_result = 'fba_sol_fin.csv', method = 'fba')     
    
    """
    
    modelDir = os.getcwd()                                                               
    cmod=cbm.CBRead.readSBML3FBC(model, modelDir)
    cmod.createGeneAssociationsFromAnnotations()

    #fva
    fva_res, names = cbm.FluxVariabilityAnalysis(cmod, optPercentage=100)
    #fba
    fba_sol = cbm.analyzeModel(cmod, return_lp_obj=True) 
    sol_dist = fba_sol.solution.get_values()

    #finding reactions that got significantly affected by perturbation
    fba_np = np.genfromtxt(pert_result, delimiter=',')
    
    #rand_np = np.genfromtxt('rand_val_test.csv', delimiter=',')
    
    rng = range(10) #defining the number of perturbations

    opt_som = []
    
    c_count = 1 #counter for the perturbed reactions
    l_count = 1 #counter for perturbation number
    
    
    #finding the significant flux changes and sorting the results => perturbed reaction, the number of perturbation, the affected reaction (significant         flux change)
    for j in range(len((fba_np))):
        for i in range(len((fba_np[0]))):

                        
            if method == 'fva':
                new_val = fba_np[j, i]
                dif_abs = abs(new_val - fva_res[i,0])
                
            if method == 'fba':
                new_val = fba_np[j, i]
                dif_abs = abs(new_val - sol_dist[i])
                
            
            #defining the tolerance for non-zero fluxe values (tol1) and zero flux values (tol2)
            tol1 = 5*abs(fva_res[i,0])/100
            tol2 = 0.000001
            if abs(fva_res[i,0]) !=0:
                tol = tol1
            else:
                tol = tol2
            
            #checking whether the flux change is larger thant the tolerance    
            if dif_abs > tol:
                opt_som.append([c_count, l_count, i+1, fba_np[j, i], fva_res[i,0], tol]) #appending the significantly affected reactions, the number of                                                                                              perturbation  and the respective flux values
        
            if i == len((fba_np[0]))-1:
                
                l_count = l_count +1
                
            if j/c_count == 10:
                
                c_count = c_count +1
                l_count = 1
        
    np.savetxt("final.csv", opt_som, delimiter=",")
    print (opt_som)
